src/UnionFind.hs

```diff
@@ -0,0 +1,201 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+-- | This module implements a persistent union-find data structure. This
+-- structure reprensents a map from equivalence classes of keys to values. The
+-- equivalence relation is irrelevant. The structure primarily provides two
+-- operations:
+--
+-- - @union@ takes two keys and merges their equivalence classes, returning a
+--   new structure. Because both keys might be associated to a value, this
+--   operation must also be told how to reconcile both values into one.
+--
+-- - @find@ takes a key and returns the value associated to the key's
+--   equivalence class. In order for the structure to remain efficient, @find@
+--   also optimises it for future calls and therefore returns a new structure.
+
+module UnionFind where
+
+import Data.Default (Default, def)
+import Data.Either (partitionEithers)
+import Data.Function ((&))
+import Data.Map (Map)
+import qualified Data.Map as Map
+import Data.Maybe (fromJust)
+import Prettyprinter
+
+data UnionFindCell key value
+  = ChildOf key
+  | Ancestor
+  | AncestorWith value
+
+-- | A persistent union-find data structure.
+--
+-- Invariants:
+--
+--   (i) There are no cycles.
+--   (ii) Keys under @ChildOf@ are present in the map.
+--
+-- The points (i) and (ii) imply that, if a key is present in the map, then
+-- there is a path made of @ChildOf@ constructors from that key to an @Ancestor@
+-- or @AncestorWith@ constructor.
+--
+newtype UnionFind key value = UnionFind {
+  getMap :: Map key (UnionFindCell key value)
+}
+
+instance Default (UnionFind key value) where
+  def = empty
+
+instance (Pretty key, Pretty value) => Pretty (UnionFind key value) where
+  pretty unionFind =
+    vsep $ map (uncurry prettyBinding) (Map.toList $ getMap unionFind)
+    where
+      prettyBinding key1 (ChildOf key2) = pretty key1 <+> "~~" <+> pretty key2
+      prettyBinding key (AncestorWith value) = pretty key <+> ":=" <+> pretty value
+
+-- | The empty union-find that does not know of any equivalences or values.
+--
+empty :: UnionFind key value
+empty = UnionFind Map.empty
+
+-- | @find key unionFind@ finds the ancestor and value associated to @key@ in
+-- the @unionFind@ structure. It returns a new union-find structure with the
+-- same semantics but optimised for future calls, the ancestor for @key@ and the
+-- found value, @Just Nothing@ if the @key@ is in the structure but without
+-- binding, or @Nothing@ if @key@ is not in the structure.
+--
+-- Prefer using @lookup@.
+--
+-- FIXME: Maybe we prefer a custom type instead of `Maybe (Maybe value)`?
+-- FIXME: Implement the optimisation for future calls
+--
+find :: Ord key =>
+  key ->
+  UnionFind key value ->
+  (UnionFind key value, key, Maybe (Maybe value))
+find key unionFind =
+  case Map.lookup key $ getMap unionFind of
+    Nothing -> (unionFind, key, Nothing)
+    Just (ChildOf key') -> find key' unionFind -- will be @Just@
```

Comment:
Should we split `find` into two functions, adding one where we know for sure that the result exists (and `error` otherwise) and that we would call from here to enforce the invariant?

Reply:
I suppose it would just give us a slightly stronger check that the invariants hold. Here, line 80, because there is a `ChildOf key'`, we know that `key'` *must be* in the map. Therefore, we know that `find key' unionFind` will evaluate to `Just`. With the current implementation, though, it could evaluate to `Nothing` without us ever noticing.

Reply:
Ah that makes sense. Yea, better guarantees can't hurt

Reply:
Done in a70eb9d.

Reply:
And this was apparently a good call because the invariant *does* get violated in some tests.